homeassistant/components/rituals_perfume_genie/__init__.py

```diff
@@ -30,23 +31,16 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
 
     try:
         account_devices = await account.get_devices()
-    except ClientConnectorError as ex:
-        raise ConfigEntryNotReady from ex
-
-    hublots = []
-    devices = {}
-    for device in account_devices:
-        hublot = device.data[HUB][HUBLOT]
-        hublots.append(hublot)
-        devices[hublot] = device
+    except aiohttp.ClientError as err:
+        raise ConfigEntryNotReady from err
 
     hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
         COORDINATORS: {},
-        DEVICES: devices,
+        DEVICES: {},
     }
 
-    for hublot in hublots:
-        device = hass.data[DOMAIN][entry.entry_id][DEVICES][hublot]
+    for device in account_devices:
+        hublot = device.data[HUB][HUBLOT]
 
         async def async_update_data():
```

Comment:
If we subclass the `DataCoordinator` we can pass the `device` to the class when instantiating it, and save the device as an instance attribute. We can avoid the closure then. Implement this method as update method:

https://github.com/home-assistant/core/blob/3a0b0380c7acbbe06f793c8b11bdc962609f0002/homeassistant/helpers/update_coordinator.py#L135-L136

Reply:
See eg:
https://github.com/home-assistant/core/blob/3a0b0380c7acbbe06f793c8b11bdc962609f0002/homeassistant/components/wled/__init__.py#L108-L137